pkg/lib/credentials.js

```diff
@@ -180,78 +180,78 @@ function Keys() {
         return cockpit.script('dir=$(dirname "$1"); test -e "$dir" || mkdir -m 700 "$dir"', [file]);
     }
 
-    function run_keygen(file, new_type, old_pass, new_pass, two_pass) {
+    async function run_keygen(file, new_type, old_pass, new_pass, two_pass) {
         const old_exps = [/.*Enter old passphrase: $/];
         const new_exps = [/.*Enter passphrase.*/, /.*Enter new passphrase.*/, /.*Enter same passphrase again: $/];
         const bad_exps = [/.*failed: passphrase is too short.*/];
 
-        return new Promise((resolve, reject) => {
-            let buffer = "";
-            let sent_new = false;
-            let failure = _("No such file or directory");
+        let buffer = "";
+        let sent_new = false;
+        let failure = _("No such file or directory");
+
+        if (new_pass !== two_pass)
+            return Promise.reject(new Error(_("The passwords do not match.")));
+
+        // Exactly one of new_type or old_pass must be given
+        console.assert((new_type == null) != (old_pass == null));
+
+        const cmd = ["ssh-keygen", "-f", file];
+        if (new_type)
+            cmd.push("-t", new_type);
+        else
+            cmd.push("-p");
+
+        await self.p_have_path;
+
+        const proc = cockpit.spawn(cmd, { pty: true, environ: ["LC_ALL=C"], err: "out", directory: self.path });
+
+        const timeout = window.setTimeout(() => {
+            failure = _("Prompting via ssh-keygen timed out");
+            proc.close("terminated");
+        }, 10 * 1000);
+
+        proc.stream(data => {
+            buffer += data;
+            if (old_pass) {
+                for (let i = 0; i < old_exps.length; i++) {
```

Comment:
I know this is existing code, but damn that's hard to read.  I feel like this would be better written as something like `if (old_exps.some(exp => exp.test(buffer)))`...

Reply:
Done.